rfc-172-graphql-for-govuk.md

```diff
@@ -0,0 +1,509 @@
+---
+status: proposed
+implementation: proposed
+status_last_reviewed: 2024-05-13
+---
+
+# GraphQL for GOV.UK
+
+Summary
+-------
+
+We propose adding a new [GraphQL][graphql] content delivery API to GOV.UK. This API would provide a read-only view of GOV.UK's content, similar to the existing [content-store][content-store] API, but supporting much more flexible queries. In particular:
+
+- Selecting only the subset of fields relevant to the client
+
+- Dynamically following links between content relevant to the client, for example on /government/ministers index page [listing the ministers, and for each minister all of the roles they currently hold](#query-to-get-the-content-on-the-ministers-index-page)
+
+- Aggregations of content, for example [listing governments](#query-to-list-governments) or [listing documents related to a topical event](#query-to-list-documents-related-to-a-topical-event)
+
+The implementation of this API would avoid many of the issues which make GOV.UK's current [publishing-api][publishing-api] / [content-store][content-store] / [search-api-v1][search-api-v1] architecture so complicated. In particular, it would avoid precomputing link expansion at the point when documents are published, and instead collect the links relevant to the client at query time. Similarly, with appropriate indexes in place, the API would cover many (or possibly all) of [the use cases we have for search-api-v1][search-api-v1-use-cases] and its associated Elasticsearch database.
```

Comment:
Is this assuming any relevancy based use cases are covered by Search API v2?

Reply:
It's handwaving a bit I think 😅  - could do with being more explicit. I think:

1) Use cases with no query, ordered by newest (e.g. latest pages in a topical event) are clear cut - should be graphql
2) Use cases with no query, ordered by popularity (e.g. /search/services landing page) are only doable if we move popularity statistics into publishing-api, because if we need to do `select ... order by ... limit ...` then the database needs to have the field we're ordering on. Personally I'd go for that, but probably a separate discussion.
3) Use cases with a query, where we're happy with keyword search (rather than semantic search) are only doable if we have some kind of free text search capability in a DB that the graphql API has access to (could still be publishing-api with postgres free text search, or it could be an OpenSearch database or something if we don't mind maintaining another DB)
4) Use cases that require semantic search / relevance based ordering should be Search API v2

Reply:
Removed the rather ambitious "(or possibly all)" parenthetical in 76181bc1cd9ef9c2f88b0ee65170a2c3c5c883e3, and quoted this thread in the commit message.